segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -263,60 +375,107 @@ public void commitRecords(Collection<SystemJournalRecord> records) throws ChunkS
         log.debug("SystemJournal[{}] Logging system log records - file={}, batch={}.", containerId, currentHandle.getChunkName(), batch);
     }
 
+    /**
+     * Generate a snapshot if required.
+     */
+    private void generateSnapshotIfRequired() {
+        synchronized (lock) {
+            if (recordsSinceSnapshot > config.getMaxJournalRecordsPerSnapshot() ||
+                currentTimeSupplier.get() - lastSavedSnapshotTime > config.getJournalSnapshotInfoUpdateFrequency().toMillis()) {
+                // Write a snapshot.
+                try (val txn = metadataStore.beginTransaction(true, getSystemSegments())) {
+                    validateAndSaveSnapshot(txn, true, config.isSelfCheckEnabled());
```

Comment:
This is performing an async operation writing to BK (and possibly other async stuff). In its current form, you are also awaiting the result of a future synchronously (blocking). This will lead to poor performance (since you're holding the lock) and possible deadlocks. 

Please rework this class so:
1. You do not need to synchronously await futures.
2. Do not do async operations or otherwise any blocking while holding locks.

Reply:
Fixed.